## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
import seaborn as sns


titanic_data = sns.load_dataset("titanic")
titanic_data.sample(5)

/Users/hacker1/Desktop/lessons/sem2py/lib/python3.14/site-packages/seaborn/utils.py:602: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df["class"] = pd.Categorical(df["class"], ["First", "Second", "Third"])
/Users/hacker1/Desktop/lessons/sem

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
414,1,3,male,44.0,0,0,7.9250,S,Third,man,True,NaN,Southampton,yes,True
431,1,3,female,NaN,1,0,16.1000,S,Third,woman,False,NaN,Southampton,yes,False
97,1,1,male,23.0,0,1,63.3583,C,First,man,True,D,Cherbourg,yes,False
802,1,1,male,11.0,1,2,120.0000,S,First,child,False,B,Southampton,yes,False
276,0,3,female,45.0,0,0,7.7500,S,Third,woman,False,NaN,Southampton,no,True


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [ ]:
empty_col = titanic_data.isnull().sum()

In [ ]:
empty_col #for checking :)

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [ ]:
m = titanic_data.shape[0]
deleted_data = titanic_data.dropna(axis="columns", thresh=(m/2) ) 

n = deleted_data.shape[1]
deleted_data = deleted_data.dropna(axis="index", thresh=(n/2))

deleted_data #result

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,NaN,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [ ]:
man_median = round(
    deleted_data[deleted_data["who"] == "man"]["age"].median()
)

woman_median = round(
    deleted_data[deleted_data["who"] == "woman"]["age"].median()
)

child_median = round(
    deleted_data[deleted_data["who"] == "child"]["age"].median()
)


deleted_data.loc[(deleted_data["who"] == "man") & (deleted_data["age"].isna()), "age" ] = man_median
deleted_data.loc[(deleted_data["who"] == "woman") & (deleted_data["age"].isna()), "age"] = woman_median 
deleted_data.loc[(deleted_data["who"] == "child") & (deleted_data["age"].isna()), "age"] = child_median 

deleted_data




0


### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [30]:
n = deleted_data.shape[1]
deleted_data = deleted_data.dropna(axis="index", thresh=(n - 1))


In [31]:
deleted_data

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,0,2,male,27.0,0,0,13.0000,S,Second,man,True,Southampton,no,True
887,1,1,female,19.0,0,0,30.0000,S,First,woman,False,Southampton,yes,True
888,0,3,female,30.0,1,2,23.4500,S,Third,woman,False,Southampton,no,False
889,1,1,male,26.0,0,0,30.0000,C,First,man,True,Cherbourg,yes,True


### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [ ]:
most_popular_town = deleted_data["embarked"].value_counts().idxmax()


for_result = {"C": "Cherbourg",
              "Q": "Queenstown",
              "S": "Southampton"}


print(for_result[most_popular_town])


Southampton


### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [47]:
n = deleted_data.shape[0]
survived_n = deleted_data["survived"].sum()
survived_percent = round(
    (survived_n / n) * 100,
    2
)
print(survived_percent)

38.25


### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [55]:
surv_emb_dict = dict()
names = ["Southampton", "Cherbourg", "Queenstown"]
for name in names:
    surv_emb_dict[name] = deleted_data[
        deleted_data["embark_town"]
        == name]["survived"].sum()
    
surv_emb_series = pd.Series(surv_emb_dict)
surv_emb_series

Southampton    217
Cherbourg       93
Queenstown      30
dtype: int64

### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [59]:
surv_classes_dict = dict()
classes = ["First", "Second", "Third"]
for class_name in classes:
    psngrs = deleted_data[deleted_data["class"] == class_name]

    all_psngrs = len(psngrs)
    survived_psngrs = psngrs["survived"].sum()
    surv_classes_dict[class_name] = round(
        (survived_psngrs / all_psngrs) * 100,
        2
    )

    
surv_classes_series = pd.Series(surv_classes_dict)
surv_classes_series

First     62.62
Second    47.28
Third     24.24
dtype: float64

### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [61]:
rich_part = deleted_data[ 
    deleted_data["fare"] >= 100
]

all_rich = len(rich_part)
survived_rich = rich_part["survived"].sum()

rich_percent = round(
    (survived_rich / all_rich) * 100, 
    2
)

print(rich_percent)


73.58


### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [ ]:
lone_childs = deleted_data[
    (deleted_data["who"] == "child") &
    (deleted_data["alone"] == True) 
]

print(
    len(lone_childs)
    )

6


Какие выводы вы можете сделать о выживших пассажирах Титаника? 

Большинство из них -- люди из первого класса. Это очевидно, так как их каюты находились наверху, ближе к шлюпкам, их оповещали первыми, сажали приоритетнее. Также вперед пропускали женщин и детей.

Тот факт, что общий процент выживших 38 -- грустно 😞🙁